### DISCUSSION TITLE
OpenAI Assistants

### DISCUSSION THREAD

**Discussion by Jaaco** [davidmigloz/langchain_dart#282] [Q&A] [answered]
I want to give my agent access to pdf files for knowledge. The way I found to do this would be to use "assistants" from OpenAI. I did not see if there is a way to access them via langchain.dart.

The "hacky" way I see to do this within the package right now would be to have the data available on the device in a .txt file. This has the disadvantage that 1. I only have access to the data from the time of compiling and 2. can't use pdfs, so I'd have to convert them to .txt files.

So my questions are:
Is using OpenAI assistants currently possible?
If not, is it on the roadmap (or even relevant to this project)?
Is there a different  / better way to achieve what I am trying to do?

Cheers :)

**Comment by davidmigloz** [comment, marked as answer]
Hey @Jaaco!

Thanks for your interest.

There are not many options to extract text from PDFs reliably in Dart, unfortunately. Some people are using [syncfusion_flutter_pdf](https://pub.dev/packages/syncfusion_flutter_pdf), but the results are not great and you may need to buy a license depending on your needs. You can check [this discussion](https://discord.com/channels/1123158322812555295/1182349217704976464) in our Discord where we discussed this topic.

Once you have the text, you can use LangChain.dart to embed the text, store it in a vector database, query it and augment it using an LLM (typical Retrieval Augmented Generation pipeline).

Regarding OpenAI Assistants, I am planning to add support for it next week. I already updated our OpenAI client to support them, now I need to implement the agent that interacts with it.

https://python.langchain.com/docs/modules/agents/agent_types/openai_assistants

**Reply by Jaaco** [reply]
That sounds really good, thanks for all the amazing work.
